### PR TITLE
Add (more generic) support for Circl signature schemes

### PR DESCRIPTION
**Pull request by bwesterb**
[Example usage](https://gist.github.com/bwesterb/621ac21e5f65ff9587ad6b00a2c41a99)

### CODE REVIEW

**Comment by claucece** [comment]
I'll check this tomorrow as well ;)

File: src/crypto/tls/auth_test.go [hunk 1]
```diff
@@ -161,7 +161,7 @@ func TestSupportedSignatureAlgorithms(t *testing.T) {
 		if sigType == 0 {
 			t.Errorf("%#04x: missing signature type", sigAlg)
 		}
-		if hash == 0 && sigAlg != Ed25519 {
+		if hash == 0 && sigAlg != Ed25519 && sigAlg != EdDilithium3 {
```
Comment (claucece) [inline src/crypto/tls/auth_test.go]: Maybe: `if hash == 0 && (sigAlg != Ed25519 || sigAlg != EdDilithium3)`?

File: src/crypto/tls/generate_cert.go [hunk 2]
```diff
@@ -28,13 +29,14 @@ import (
...
-	ecdsaCurve = flag.String("ecdsa-curve", "", "ECDSA curve to use to generate a key. Valid values are P224, P256 (recommended), P384, P521")
-	ed25519Key = flag.Bool("ed25519", false, "Generate an Ed25519 key")
+	host            = flag.String("host", "", "Comma-separated hostnames and IPs to generate a certificate for")
+	validFrom       = flag.String("start-date", "", "Creation date formatted as Jan 1 15:04:05 2011")
+	validFor        = flag.Duration("duration", 365*24*time.Hour, "Duration that certificate is valid for")
+	isCA            = flag.Bool("ca", false, "whether this cert should be its own Certificate Authority")
+	rsaBits         = flag.Int("rsa-bits", 2048, "Size of RSA key to generate. Ignored if --ecdsa-curve is set")
+	ecdsaCurve      = flag.String("ecdsa-curve", "", "ECDSA curve to use to generate a key. Valid values are P224, P256 (recommended), P384, P521")
+	ed25519Key      = flag.Bool("ed25519", false, "Generate an Ed25519 key")
+	eddilithium3Key = flag.Bool("ed25519-dilithium3", false, "Generate an Ed25519-Dilithium3 key")
```
Comment (claucece) [inline src/crypto/tls/generate_cert.go]: As we will probably change this file, at some point I was thinking that we can change these flags logic to something nicer, as adding more and more lines here can get chaotic.. what do you think?

File: src/crypto/x509/pkcs8.go [hunk 3]
```diff
@@ -74,6 +76,20 @@ func ParsePKCS8PrivateKey(der []byte) (key interface{}, err error) {
 		}
 		return ed25519.NewKeyFromSeed(curvePrivateKey), nil
 
+	case privKey.Algo.Algorithm.Equal(oidPublicKeyEdDilithium3):
+		if l := len(privKey.Algo.Parameters.FullBytes); l != 0 {
+			return nil, errors.New("x509: invalid Ed25519-Dilithium3 private key parameters")
+		}
+		var packedSk []byte
```
Comment (claucece) [inline src/crypto/x509/pkcs8.go:87]: Let's change the 'packedSk' name to 'curvePrivateKey' ;)

File: src/crypto/x509/x509.go [hunk 4]
```diff
@@ -855,7 +868,7 @@ func checkSignature(algo SignatureAlgorithm, signed, signature []byte, publicKey
 
 	switch hashType {
 	case crypto.Hash(0):
-		if pubKeyAlgo != Ed25519 {
+		if pubKeyAlgo != Ed25519 && pubKeyAlgo != EdDilithium3 {
```
Comment (claucece) [inline src/crypto/x509/x509.go]: Perhaps: `pubKeyAlgo != Ed25519 || pubKeyAlgo != EdDilithium3`

File: src/crypto/x509/x509.go [hunk 5]
```diff
@@ -2020,7 +2054,7 @@ func signingParamsForPublicKey(pub interface{}, requestedSigAlgo SignatureAlgori
 				return
 			}
 			sigAlgo.Algorithm, hashFunc = details.oid, details.hash
-			if hashFunc == 0 && pubType != Ed25519 {
+			if hashFunc == 0 && (pubType != Ed25519 && pubType != EdDilithium3) {
```
Comment (claucece) [inline src/crypto/x509/x509.go]: The same.. why not use || here?

File: src/crypto/tls/auth.go [hunk 6]
```diff
@@ -123,10 +125,15 @@ func typeAndHashFromSignatureScheme(signatureAlgorithm SignatureScheme) (sigType
...
-		sigType = signatureEdDilithium3
 	default:
-		return 0, 0, fmt.Errorf("unsupported signature algorithm: %#04x", signatureAlgorithm)
+		scheme := circlSign.SchemeByTLSIdentifier(uint(signatureAlgorithm))
+		if scheme == nil {
+			return 0, 0, fmt.Errorf("unsupported signature algorithm: %#04x", signatureAlgorithm)
+		}
+		sigType = sigTypeByCirclScheme(scheme)
+		if sigType == 0 {
+			return 0, 0, errors.New("circl scheme not supported")
```
Comment (claucece) [inline src/crypto/tls/auth.go:135]: Probably also use `fmt.Errorf` and list the scheme..

File: src/crypto/tls/auth.go [hunk 7]
```diff
@@ -139,10 +146,12 @@ func typeAndHashFromSignatureScheme(signatureAlgorithm SignatureScheme) (sigType
...
 		hash = directSigning
-	case EdDilithium3:
-		hash = directSigning
 	default:
-		return 0, 0, fmt.Errorf("unsupported signature algorithm: %#04x", signatureAlgorithm)
+		scheme := circlSign.SchemeByTLSIdentifier(uint(signatureAlgorithm))
+		if scheme == nil {
+			return 0, 0, fmt.Errorf("unsupported signature algorithm: %#04x", signatureAlgorithm)
+		}
+		hash = directSigning
```
Comment (claucece) [inline src/crypto/tls/auth.go:154]: All the schemes on circle will have directSigning, right?

File: src/crypto/tls/auth_test.go [hunk 8]
```diff
@@ -161,7 +163,7 @@ func TestSupportedSignatureAlgorithms(t *testing.T) {
 		if sigType == 0 {
 			t.Errorf("%#04x: missing signature type", sigAlg)
 		}
-		if hash == 0 && sigAlg != Ed25519 && sigAlg != EdDilithium3 {
+		if hash == 0 && sigAlg != Ed25519 && circlSign.SchemeByTLSIdentifier(uint(sigAlg)) == nil {
```
Comment (claucece) [inline src/crypto/tls/auth_test.go:166]: Mmm.. this does not read nicely...

File: src/crypto/tls/generate_cert.go [hunk 9]
```diff
@@ -29,14 +29,14 @@ import (
...
-	ed25519Key      = flag.Bool("ed25519", false, "Generate an Ed25519 key")
-	eddilithium3Key = flag.Bool("ed25519-dilithium3", false, "Generate an Ed25519-Dilithium3 key")
+	host       = flag.String("host", "", "Comma-separated hostnames and IPs to generate a certificate for")
+	validFrom  = flag.String("start-date", "", "Creation date formatted as Jan 1 15:04:05 2011")
+	validFor   = flag.Duration("duration", 365*24*time.Hour, "Duration that certificate is valid for")
+	isCA       = flag.Bool("ca", false, "whether this cert should be its own Certificate Authority")
+	rsaBits    = flag.Int("rsa-bits", 2048, "Size of RSA key to generate. Ignored if --ecdsa-curve is set")
+	ecdsaCurve = flag.String("ecdsa-curve", "", "ECDSA curve to use to generate a key. Valid values are P224, P256 (recommended), P384, P521")
+	ed25519Key = flag.Bool("ed25519", false, "Generate an Ed25519 key")
+	circlKey   = flag.String("circl", false, "Generate a key supported by Circl")
```
Comment (claucece) [inline src/crypto/tls/generate_cert.go:39]: Should we use as value false here?

File: src/crypto/tls/handshake_client.go [hunk 10]
```diff
@@ -828,7 +829,7 @@ func (c *Conn) verifyServerCertificate(certificates [][]byte) error {
 	}
 
 	switch certs[0].PublicKey.(type) {
-	case *rsa.PublicKey, *ecdsa.PublicKey, ed25519.PublicKey, *eddilithium3.PublicKey:
+	case *rsa.PublicKey, *ecdsa.PublicKey, ed25519.PublicKey, circlSign.PublicKey:
```
Comment (claucece) [inline src/crypto/tls/handshake_client.go:832]: Why not a pointer anymore?

Comment (bwesterb) [reply src/crypto/tls/handshake_client.go:832]: `circSign.PublicKey` is an interface.

File: src/crypto/tls/generate_cert.go (hunk 9 above)
Comment (bwesterb) [reply src/crypto/tls/generate_cert.go:39]: Should've been `""`, thanks.

File: src/crypto/tls/auth.go (hunk 7 above)
Comment (bwesterb) [reply src/crypto/tls/auth.go:154]: Yes, for the moment.

File: src/crypto/tls/auth.go (hunk 6 above)
Comment (bwesterb) [reply src/crypto/tls/auth.go:135]: Done

File: src/crypto/x509/x509.go (hunk 4 above)
Comment (bwesterb) [reply src/crypto/x509/x509.go]: (You've commented on an outdated commit.  In the latest version EdDilithium3 does not appear directly but rather the scheme is used.)

Direct signing should only be used with Ed25519 or EdDilithium3.  So it's ok when `pubKeyAlgo == Ed25519 || pubKeyAlgo == EdDiltihium3`.  Negating and pulling through the negation turns the or into an and.

File: src/crypto/x509/x509.go (hunk 5 above)
Comment (bwesterb) [reply src/crypto/x509/x509.go]: See other thread.

File: src/crypto/tls/auth_test.go (hunk 1 above)
Comment (bwesterb) [reply src/crypto/tls/auth_test.go]: See other thread.

File: src/crypto/tls/generate_cert.go (hunk 2 above)
Comment (bwesterb) [reply src/crypto/tls/generate_cert.go]: Yeah, in the non-outdated commit it has a string flag to identify the scheme.

File: src/crypto/tls/auth_test.go (hunk 8 above)
Comment (bwesterb) [reply src/crypto/tls/auth_test.go:166]: What is your counter-proposal?

File: src/crypto/x509/pkcs8.go (hunk 3 above)
Comment (bwesterb) [reply src/crypto/x509/pkcs8.go:87]: But it does not necessarily have anything to do with a curve, does it?

Comment (bwesterb) [reply src/crypto/x509/pkcs8.go:87]: (I'm reading your comment as if you've made it on the latest version with he generic code.)

**Comment by bwesterb** [comment]
Updated with API changes of @armfazh

File: src/crypto/x509/pkcs8.go (hunk 3 above)
Comment (claucece) [reply src/crypto/x509/pkcs8.go:88]: No, it is more around following the names that are already there.

File: src/crypto/x509/x509.go (hunk 4 above)
Comment (claucece) [reply src/crypto/x509/x509.go]: Gotcha.

**Comment by claucece** [comment]
Looking good for me @bwesterb ! What is it missing?

**Comment by bwesterb** [comment]
> Looking good for me @bwesterb ! What is it missing?

I'm waiting for https://github.com/cloudflare/circl/pull/143 to be merged.

**Pull request merged**